### PR TITLE
Popcount32 functional correctness and (non)effects characterization

### PR DESCRIPTION
**Pull request by shigoel**
### Description:

Prove that the popcount32 program correctly computes the bitcount of its input, and prove that certain state components were left unmodified by this program.

### Testing:

`make all` ran locally.

### License:

By submitting this pull request, I confirm that my contribution is
made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by alexkeizer — COMMENTED** [review]
@shigoel the non-effects proof now goes through fully automatically. The key was setting the `contextual` of simp, so that the non-equality hypotheses were automatically added to the simpset.

File: Proofs/Popcount32.lean [hunk 1]
```diff
@@ -68,44 +71,57 @@ def popcount32_program : Program :=
...
---        clear h_step_21 h_step_22 h_step_23 h_step_24;
---        clear h_step_25 h_step_26)
---   -- Final Steps
---   unfold run at h_run
---   subst s_final
---   unfold popcount32_spec
---   sorry
+  -- Final Steps
+  -- Split all the Ands in the conclusion.
+  repeat (any_goals apply And.intro)
```
Comment (alexkeizer) [inline Proofs/Popcount32.lean:93]: `repeat (any_goals _)` looks a bit funny, is `repeat` by itself not enough?

File: Proofs/Popcount32.lean [hunk 2]
```diff
@@ -68,44 +71,57 @@ def popcount32_program : Program :=
...
+  · simp only [popcount32_spec,
+               fst_AddWithCarry_eq_add,
+               fst_AddWithCarry_eq_sub_neg]
+    repeat (simp only [popcount32_spec_rec])
```
Comment (alexkeizer) [inline Proofs/Popcount32.lean:97]: Here, too, `repeat (simp _)` looks weird. We only have one goal, right? So `simp` should already repeatedly apply the rewrite

**Review by alexkeizer — COMMENTED** [review]
LGTM, modulo two minor comments

File: Proofs/Popcount32.lean (hunk 1 above)
Comment (alexkeizer) [reply Proofs/Popcount32.lean:93]: @bollu told me that, no, `repeat` will only repeat on the main goal.

However, `repeat'` does work on all goals, and should work here

Comment (shigoel) [reply Proofs/Popcount32.lean:93]: Hah, thanks!

**Pull request merged**